### PR TITLE
Add XBPM feedback for i03

### PR DESCRIPTION
**Pull request by DominicOram**
Merge https://github.com/DiamondLightSource/dodal/pull/142 first.

Needed for https://github.com/DiamondLightSource/hyperion/issues/796

### CODE REVIEW

File: src/dodal/devices/i04/transfocator.py
Comment (Tom-Willemsen) [inline src/dodal/devices/i04/transfocator.py:24]:
```suggestion
    def polling_wait_on_start_rbv(self, for_value):
```

Unless it's meant to be `stat_rbv`, in which case the field name (and PV name) are likely wrong

File: src/dodal/devices/i04/transfocator.py
Comment (Tom-Willemsen) [inline src/dodal/devices/i04/transfocator.py:32]: This might take much longer than `TIMEOUT` seconds, especially if `.get()` is slow for any reason.

How about something like:

```python
def polling_wait_on_stat_rbv(self, for_value):
    start_time = time.time()
    while time.time() < start_time + self.TIMEOUT:
        RBV_value = self.start_rbv.get()
        if RBV_value == for_value:
                return
        sleep(self._POLLING_WAIT)

    # last try
    if self.start_rbv.get() != for_value:
        raise TimeoutError()
```

File: src/dodal/devices/i04/transfocator.py
Comment (Tom-Willemsen) [inline src/dodal/devices/i04/transfocator.py:39]: Do we want to compare with a floating-point tolerance here?

File: src/dodal/devices/i04/transfocator.py
Comment (Tom-Willemsen) [inline src/dodal/devices/i04/transfocator.py:42]: As someone who isn't an expert on this device this seems a bit surprising. Can we justify it with a comment?

File: src/dodal/devices/i04/transfocator.py
Comment (Tom-Willemsen) [inline src/dodal/devices/i04/transfocator.py:58]: There's a load of fiddly stuff going on in this method which seems non-standard to me. Can we add a docstring on the class explaining what the device physically does and how this code achieves that goal?

File: src/dodal/devices/xbpm_feedback.py
Comment (Tom-Willemsen) [inline src/dodal/devices/xbpm_feedback.py:9]: Good that we have a comment but can we expand it a little for people who aren't so familiar with the device?

Something along the lines of:

> We need to wait for `<something>` to go high, wait for `<some>` seconds, and then wait for it to go low again to ensure that `<some process>` has completed succesfully.

File: tests/devices/i04/test_transfocator.py
Comment (Tom-Willemsen) [inline tests/devices/i04/test_transfocator.py:51]: This looks like it is a race condition against the `poll_period`? Is there not a way we can write this test with mocking etc to avoid these sleep calls?

I'm thinking you should be able to set up the mock to return `1` on the first call to `get()` and `0` subsequently.

File: pyproject.toml
Comment (DominicOram) [reply pyproject.toml:17]: I discussed this with the core `ophyd` developers yesterday. I think @callumforrester was proposing something similar to your suggestion but I would rather not fork if we don't have to. The general consensus in the meeting was that individual facilities should be freer to merge things into `ophyd` without other facilities and we came up with a list of named maintainers on ophyd for bigger changes. I think we try with that for now and see how far we get.

File: src/dodal/devices/i04/transfocator.py
Comment (DominicOram) [reply src/dodal/devices/i04/transfocator.py:39]: Scratch that, it stupidly is a float

File: tests/devices/i04/test_transfocator.py
Comment (DominicOram) [reply tests/devices/i04/test_transfocator.py:51]: Thanks, I've removed a couple of the sleeps, I think one might have to remain though.

File: pyproject.toml
Comment (Tom-Willemsen) [reply pyproject.toml:27]: Ah ok didn't realise our auto-formatter touched `.toml`... In that case I agree with you.

File: src/dodal/devices/i04/transfocator.py
Comment (Tom-Willemsen) [inline src/dodal/devices/i04/transfocator.py:71]:
```suggestion
            status.set_finished()
```

Otherwise I think this could deadlock if it gets into this subscriber but doesn't match the `not isclose` condition, since the status would never be set to finished?

File: src/dodal/devices/i04/transfocator.py
Comment (Tom-Willemsen) [inline src/dodal/devices/i04/transfocator.py:49]: nit:

```suggestion
        """To set the beamsize on the transfocator we must:
```

**Review by Tom-Willemsen — COMMENTED** [review]
Sorry, just one more substantive comment I noticed...

**Comment by DominicOram** [comment]
From https://jira.diamond.ac.uk/browse/I03-870 it looks like the stability check is no longer needed in `ophyd`.

File: src/dodal/devices/i04/transfocator.py
Comment (DominicOram) [reply src/dodal/devices/i04/transfocator.py:71]: I don't think we'll end up with a deadlock. The `status` will eventually just timeout in error. It's reasonable to ask if we want that to happen though. I think you're right that actually if the calculation gives the same number of lenses as before we should just immediately set finished. My worry here is that we're assuming the IOC behaves well and doesn't send us a  repeat first, I hope that's a reasonable assumption.

Comment (Tom-Willemsen) [reply src/dodal/devices/i04/transfocator.py]: Ah ok and the timeout doesn't cause an error? Fair enough, I assumed that if it timed out that would cause the scan to fail (or if there wasn't a timeout that it would deadlock).

Comment (DominicOram) [reply src/dodal/devices/i04/transfocator.py]: Sorry, you're right it will cause an error. In which case this is required

File: src/dodal/devices/xbpm_feedback.py
Comment (Tom-Willemsen) [inline src/dodal/devices/xbpm_feedback.py:22]: [nit, ignorable]

> [Signature should be `f(*, old_value, value, **kwargs)`](https://blueskyproject.io/ophyd/user_v1/generated/ophyd.status.SubscriptionStatus.html#ophyd.status.SubscriptionStatus)

I know this is technically correct but maybe it would be clearer to use the exact signature advised in the ophyd docs?

**Review by Tom-Willemsen — APPROVED** [review]
one minor suggestion, feel free to ignore and merge this if you like. Otherwise LGTM.

**Review by Tom-Willemsen — APPROVED** [review]

**Pull request merged**
